File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    List<List<Integer>> ans = new ArrayList<>();
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();
        if(root == null){
            return ans;
        }
        Queue<TreeNode> nodes = new LinkedList<>();
        nodes.add(root);
        int size = nodes.size();
        while(!nodes.isEmpty()){
            List<Integer> inner = new ArrayList<>();
            for(int i = 0; i < size; i++){
                TreeNode curr = nodes.poll();
                if(curr == null){
                    continue;
                }
                inner.add(curr.val);
                nodes.add(curr.left);
                nodes.add(curr.right);
            }
            if(inner.size() > 0){
                ans.add(inner);
            }
            size = nodes.size();
        }
        return ans;
    }
}

File: Data Structures & Algorithms/lowest-common-ancestor-of-a-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if not root or root == p or root == q:
            return root
        
        # will be none if both targets aren't found
        left = self.lowestCommonAncestor(root.left, p, q)

        # will be none if both targets aren't found
        right = self.lowestCommonAncestor(root.right, p, q)

        # found in both subtrees, this is the lca
        if left and right:
            return root
        
        # if we only find in one subtree return that one
        return left if left else right
        